core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java

```diff
@@ -0,0 +1,215 @@
+                    key -> key.startsWith(parentPath));
+            if (!existsTargetFiles) {
+                context.key.cancel();
+            }
+            stopFutureIfPossible();
+        }
+    }
+
+    private void startFutureIfPossible() {
+        if (!isRunning() && !ctxRegistry.isEmpty()) {
+            future = CompletableFuture.runAsync(new PropertiesFileWatcherRunnable(), executor);
+        }
+    }
+
+    private void stopFutureIfPossible() {
+        checkState(future != null, "tried to stop null future");
+        if (isRunning() && ctxRegistry.isEmpty()) {
+            future.cancel(true);
+        }
+    }
+
+    private static Path getRealPath(Path path) {
+        try {
+            return path.toRealPath();
+        } catch (IOException e) {
+            throw new IllegalArgumentException("failed to locate file " + path, e);
+        }
+    }
+
+    /**
+     * Check if future for {@code WatchService} is running.
+     * @return true if future is running
+     */
+    @VisibleForTesting
+    boolean isRunning() {
+        return future != null && !future.isDone();
+    }
+
+    /**
+     * Close watch service, future, and clear registry.
+     * @throws Exception may be thrown if {@code WatchService} is already closed
+     */
+    @Override
+    public void close() throws Exception {
+        ctxRegistry.clear();
+        stopFutureIfPossible();
+        watchService.close();
+    }
+
+    private class PropertiesFileWatcherRunnable implements Runnable {
+        @Override
+        public void run() {
+            try {
+                WatchKey key;
+                while ((key = watchService.take()) != null) {
+                    for (WatchEvent<?> event : key.pollEvents()) {
+                        @SuppressWarnings("unchecked")
+                        final Path watchedPath = ((Path) key.watchable())
+                                .resolve(((WatchEvent<Path>) event).context());
+                        final Path realPath;
+                        try {
+                            realPath = watchedPath.toRealPath();
+                        } catch (IOException e) {
+                            logger.warn("skipping unable to get real path for {}", watchedPath);
+                            continue;
+                        }
+
+                        if (event.kind().equals(ENTRY_MODIFY) || event.kind().equals(ENTRY_CREATE)) {
+                            if (ctxRegistry.keySet().contains(realPath)) {
+                                try {
+                                    ctxRegistry.get(realPath).reloader.run();
+                                } catch (Exception e) {
+                                    logger.warn("unexpected error from listener: {} ", realPath, e);
+                                }
+                            }
+                        } else if (event.kind().equals(OVERFLOW)) {
+                            logger.info("failed to reload file: {}", realPath);
```

Comment:
I'm not sure which is correct, but maybe we can try reloading in this case as well?